File: src/lib/stores/history.js
import { writable } from 'svelte/store'
import { invoke } from '@tauri-apps/api/core'

function createHistoryStore() {
	const { subscribe, set, update } = writable([])

	return {
		subscribe,

		async search(query) {
			try {
				const results = await invoke('history_search', { query, limit: 50 })
				set(results)
				return results
			} catch (e) {
				console.error('Failed to search history:', e)
				return []
			}
		},

		async getRecent(limit = 100) {
			try {
				const results = await invoke('history_get_recent', { limit })
				set(results)
				return results
			} catch (e) {
				console.error('Failed to get recent history:', e)
				return []
			}
		},

		async delete(id) {
			try {
				await invoke('history_delete', { id })
				update((items) => items.filter((i) => i.id !== id))
			} catch (e) {
				console.error('Failed to delete history entry:', e)
			}
		},

		async clear(timeframe) {
			try {
				await invoke('history_clear', { timeframe })
				if (timeframe === 'all') {
					set([])
				} else {
					// Reload to reflect cleared items
					const results = await invoke('history_get_recent', { limit: 100 })
					set(results)
				}
			} catch (e) {
				console.error('Failed to clear history:', e)
			}
		},

		async suggest(query, limit = 8) {
			try {
				return await invoke('history_search', { query, limit })
			} catch (e) {
				return []
			}
		},
	}
}

export const history = createHistoryStore()

File: src-tauri/src/commands/history.rs
use tauri::{command, AppHandle, Manager};

use crate::storage::database::Database;
use crate::storage::history::HistoryEntry;

/// Search history by URL or title
#[command]
pub fn history_search(
	app: AppHandle,
	query: String,
	limit: Option<i64>,
) -> Result<Vec<HistoryEntry>, String> {
	let db = app.state::<Database>();
	db.history_search(&query, limit.unwrap_or(50))
}

/// Get recent history entries
#[command]
pub fn history_get_recent(
	app: AppHandle,
	limit: Option<i64>,
) -> Result<Vec<HistoryEntry>, String> {
	let db = app.state::<Database>();
	db.history_get_recent(limit.unwrap_or(100))
}

/// Delete a single history entry
#[command]
pub fn history_delete(app: AppHandle, id: String) -> Result<(), String> {
	let db = app.state::<Database>();
	db.history_delete(&id)
}

/// Clear history by timeframe (hour, day, week, all)
#[command]
pub fn history_clear(app: AppHandle, timeframe: String) -> Result<(), String> {
	let db = app.state::<Database>();
	db.history_clear(&timeframe)
}
